oracle/src/mevSender.js

```diff
@@ -0,0 +1,159 @@
+require('../env')
+const path = require('path')
+const BigNumber = require('bignumber.js')
+const { redis } = require('./services/redisClient')
+const logger = require('./services/logger')
+const { sendTx } = require('./tx/sendTx')
+const { getNonce, getChainId, getBlock } = require('./tx/web3')
+const { addExtraGas, checkHTTPS, watchdog } = require('./utils/utils')
+const { EXIT_CODES, EXTRA_GAS_PERCENTAGE, MAX_GAS_LIMIT } = require('./utils/constants')
+const { estimateProfit } = require('./events/processAMBCollectedSignaturesMEV')
+
+if (process.argv.length < 3) {
+  logger.error('Please check the number of arguments, config file was not provided')
+  process.exit(EXIT_CODES.GENERAL_ERROR)
+}
+
+const config = require(path.join('../config/', process.argv[2]))
+
+const { web3, mevForeign, validatorAddress } = config
+
+let chainId = 0
+let flashbotsProvider
+
+async function initialize() {
+  try {
+    const checkHttps = checkHTTPS(process.env.ORACLE_ALLOW_HTTP_FOR_RPC, logger)
+
+    web3.currentProvider.urls.forEach(checkHttps(config.id))
+
+    chainId = await getChainId(web3)
+    flashbotsProvider = await mevForeign.getFlashbotsProvider(chainId)
+    return runMain()
+  } catch (e) {
+    logger.error(e.message)
+    process.exit(EXIT_CODES.GENERAL_ERROR)
+  }
+}
+
+async function runMain() {
+  try {
+    if (redis.status === 'ready') {
+      if (config.maxProcessingTime) {
+        await watchdog(main, config.maxProcessingTime, () => {
+          logger.fatal('Max processing time reached')
+          process.exit(EXIT_CODES.MAX_TIME_REACHED)
+        })
+      } else {
+        await main()
+      }
+    }
+  } catch (e) {
+    logger.error(e)
+  }
+
+  setTimeout(runMain, config.pollingInterval)
+}
+
+async function main() {
+  try {
+    const jobs = Object.values(await redis.hgetall(config.mevJobsRedisKey)).map(JSON.parse)
+    const totalJobs = jobs.length
+
+    if (totalJobs === 0) {
+      logger.info('Nothing to process')
+      return
+    }
+
+    const { baseFeePerGas: pendingBaseFee, number: pendingBlockNumber } = await getBlock(web3, 'pending')
+    const bestJob = pickBestJob(jobs, pendingBaseFee)
+
+    if (!bestJob) {
+      logger.info({ totalJobs, pendingBaseFee }, 'No suitable job was found, waiting for a lower gas price')
+      return
+    }
+
+    const jobLogger = logger.child({ eventTransactionHash: bestJob.transactionReference })
+
+    const maxProfit = await estimateProfit(
+      mevForeign.contract,
+      mevForeign.minGasPrice,
+      bestJob.executeData,
+      bestJob.value
+    )
+
+    if (maxProfit === '0') {
+      jobLogger.info(`No MEV opportunity found when testing with min gas price ${mevForeign.minGasPrice}, removing job`)
```

Comment:
In my tests I've got the following behavior:
```
Feb  6 16:13:29 amb-bsc oracle_bridge_alpha_mevSender_1/6a29bca10ef1[2320]: {"level":30,"time":1644164009281,"msg":"Sending MEV bundles","validator":"0xACeFD9907F5D4d7Fc387193414733a111863c8D6","name":"mev-sender-foreign","eventTransactionHash":"0x2b6c8df5db39f90c0d23ad0512e7482bc15a445f3315bbddccb12993c0eb9554","nonce":4431,"fromBlock":15023194,"toBlock":15023198,"v":1}
Feb  6 16:13:29 amb-bsc oracle_bridge_alpha_mevSender_1/6a29bca10ef1[2320]: {"level":30,"time":1644164009775,"msg":"Tx generated 0xfdeec6ee313005977991457fe00b24e0b8ed8f8f0dbe89487b1f5595c4ddb1fc for event Tx 0x2b6c8df5db39f90c0d23ad0512e7482bc15a445f3315bbddccb12993c0eb9554","validator":"0xACeFD9907F5D4d7Fc387193414733a111863c8D6","name":"mev-sender-foreign","eventTransactionHash":"0x2b6c8df5db39f90c0d23ad0512e7482bc15a445f3315bbddccb12993c0eb9554","txHash":"0xfdeec6ee313005977991457fe00b24e0b8ed8f8f0dbe89487b1f5595c4ddb1fc","v":1}
Feb  6 16:13:45 amb-bsc oracle_bridge_alpha_mevSender_1/6a29bca10ef1[2320]: {"level":40,"time":1644164025014,"msg":"Gas price API is not configured, will fallback to the contract-supplied gas price","validator":"0xACeFD9907F5D4d7Fc387193414733a111863c8D6","name":"mev-sender-foreign","module":"gasPrice","chainId":"foreign","v":1}
Feb  6 16:13:45 amb-bsc oracle_bridge_alpha_mevSender_1/6a29bca10ef1[2320]: {"level":30,"time":1644164025714,"msg":"No MEV opportunity found when testing with min gas price 5000000000, removing job","validator":"0xACeFD9907F5D4d7Fc387193414733a111863c8D6","name":"mev-sender-foreign","eventTransactionHash":"0x2b6c8df5db39f90c0d23ad0512e7482bc15a445f3315bbddccb12993c0eb9554","v":1}
```
As you can see it is only 16 seconds between two actions: the transaction was sent, the corresponding job was removed. Isn't it too fast? May we have larger delay for checks whether transactions was executed successfully? On the Ethereum Mainnet it could take minutes before the transaction is included into the chain so the sender could send several transactions for the same job in short period of time.

Reply:
Well, the job is not removed immediately after it was sent to the Flashbots RPC. Indeed, there is no guarantee that it will be included at all, since we send request only for a few upcoming blocks.

After the job tx was sent to the RPC, the oracle will then again reinspect it on the next iteration, and will remove it only if it is no longer profitable (e.g. if it was already executed by oracle/someone else). If the transaction is still profitable, it will continue to send it, until it become non-profitable (e.g. when message was successfully executed).

Reply:
Ok. Understood. So does it make sense to make the polling interval independent from ORACLE_FOREIGN_RPC_POLLING_INTERVAL? So, we can control how fast the sender re-estimates the transaction?

Reply:
Well, since in the mevSender, ORACLE_FOREIGN_RPC_POLLING_INTERVAL is not used for anything else, there is no need to introduce another variable.

For example, we can `ORACLE_FOREIGN_RPC_POLLING_INTERVAL=15000` in the watcher and `ORACLE_FOREIGN_RPC_POLLING_INTERVAL=70000` in the sender.

Also, note that since sender will send MEV bundles for 5 upcoming blocks by default, there is no reason to set polling interval lower than `5*avg_block_time ~= 70 seconds`.

Reply:
Got you. In this case consider to have `ORACLE_FOREIGN_RPC_POLLING_INTERVAL` configured for mev-sender in `oracle/docker-compose-helpers.yml` with the corresponding comment. I would also say that we need to re-define `ORACLE_HOME_START_BLOCK` in the docker-compose file as a start point for watching the messages to execute.